verification/src/error.rs

```diff
@@ -96,119 +173,213 @@ pub enum CellbaseError {
+    TooOldBlockTime { min: u64, actual: u64 },
+
+    #[fail(display = "TooNewBlockTime{{max: {}, actual: {}}}", max, actual)]
+    TooNewBlockTime { max: u64, actual: u64 },
 }
 
-#[derive(Debug, PartialEq, Clone, Copy, Eq)]
+#[derive(Fail, Debug, PartialEq, Eq, Clone)]
+#[fail(display = "expected: {}, actual: {}", expected, actual)]
 pub struct NumberError {
     pub expected: u64,
     pub actual: u64,
 }
 
-#[derive(Debug, PartialEq, Clone, Eq)]
+#[derive(Fail, Debug, PartialEq, Eq, Clone)]
 pub enum EpochError {
-    DifficultyMismatch { expected: U256, actual: U256 },
-    NumberMismatch { expected: u64, actual: u64 },
-    AncestorNotFound,
-}
+    // NOTE: the original name is DifficultyMismatch
+    #[fail(
+        display = "UnmatchedDifficulty{{expected: {}, actual: {}}}",
+        expected, actual
+    )]
+    UnmatchedDifficulty { expected: Byte32, actual: Byte32 },
 
-#[derive(Debug, PartialEq, Clone, Copy, Eq)]
-pub enum TransactionError {
-    /// Occur output's bytes_len exceed capacity
-    CapacityOverflow,
-    /// In a single output cell, the capacity is not enough to hold the cell serialized size
-    InsufficientCellCapacity,
-    DuplicateDeps,
-    Empty,
-    /// Sum of all outputs capacity exceed sum of all inputs in the transaction
-    OutputsSumOverflow,
-    InvalidScript,
-    ScriptFailure(ScriptError),
-    InvalidSignature,
-    Version,
-    /// Tx not satisfied since condition
-    Immature,
-    /// Invalid Since flags
-    InvalidSince,
-    CellbaseImmaturity,
-    ExceededMaximumBlockBytes,
-    OutputsDataLengthMismatch,
-}
-
-impl StdError for TransactionError {}
-
-impl fmt::Display for TransactionError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        fmt::Debug::fmt(&self, f)
-    }
+    // NOTE: the original name is NumberMismatch
+    #[fail(
+        display = "UnmatchedNumber{{expected: {}, actual: {}}}",
+        expected, actual
+    )]
+    UnmatchedNumber { expected: u64, actual: u64 },
+
+    // NOTE: the original name is AncestorNotFound
+    #[fail(display = "MissingAncestor")]
+    MissingAncestor,
 }
 
 impl TransactionError {
-    /// Transaction error may be caused by different tip between peers if this method return false,
-    /// Otherwise we consider the Bad Tx is constructed intendedly.
-    pub fn is_bad_tx(self) -> bool {
-        use TransactionError::*;
+    pub fn is_malformed_tx(&self) -> bool {
         match self {
-            CapacityOverflow | DuplicateDeps | Empty | OutputsSumOverflow | InvalidScript
-            | ScriptFailure(_) | InvalidSignature | InvalidSince => true,
-            _ => false,
+            TransactionError::OutputOverflowCapacity
+            | TransactionError::DuplicatedDeps
+            | TransactionError::MissingInputsOrOutputs
+            | TransactionError::OccupiedOverflowCapacity
+            | TransactionError::InvalidSinceFormat
+            | TransactionError::TooLargeSize
```

Comment:
`OccupiedOverflowCapacity`, `UnmatchedOutputsDataLength`, `UnmatchedOutputsDataHashes` are `is_malformed_tx`, before they are not. But I think they should be.

But where is the `InvalidSignature` and `CapacityOverflow`?

Reply:
`InvalidSignature` has been never used.

`CapacityOverflow` added.